### PR TITLE
Invalid combination of arguments to BinnedTimeSeries raises an exception

### LINKED ISSUE #11262: Invalid combination of arguments to BinnedTimeSeries should raise a warning
### Description

I am working on a PR for providing custom binning to `aggregate_downsample` #8584. This led me to test arguments to BinnedTimeSeries, and it seems that when I pass `time_bin_start` as an array and `time_bin_end` as a scalar (uneven contiguous bins) along with an `n_bins` argument, no warning/exception is provided. The functionality is correct, so I wanted to make sure if a warning should be expected here.

### Steps to Reproduce

1. Normal Scenario
```
>>> start = Time([1, 2, 3], format='cxcsec')
>>> end = Time(5, format='cxcsec')
>>> binned = BinnedTimeSeries(time_bin_start=start, time_bin_end=end)
>>> binned
<BinnedTimeSeries length=3>
time_bin_start   time_bin_size   
                       s         
    object          float64      
-------------- ------------------
           1.0 1.0000000000065512
           2.0 0.9999999999969589
           3.0   2.00000000000351
```

2. Unexpected Scenario
```
>>> binned = BinnedTimeSeries(time_bin_start=start, time_bin_end=end, n_bins=10)
>>> binned
<BinnedTimeSeries length=3>
time_bin_start   time_bin_size   
                       s         
    object          float64      
-------------- ------------------
           1.0 1.0000000000065512
           2.0 0.9999999999969589
           3.0   2.00000000000351

```

### Expected behavior
A warning/exception telling the user that `n_bins` is not valid in this scenario.

### System Details
I am using the latest astropy dev version

```
>>> import astropy
>>> astropy.__version__
'4.3.dev434+gf58e9673d.d20210121'
```

### PR DESCRIPTION
**Pull request by bhavyakh**
A warning is raised when n_bins is passed as an argument while any of the parameters 'time_bin_start' or 'time_bin_end' is not scalar.

Fixes #11262

### CODE REVIEW

File: astropy/timeseries/binned.py
Comment (aaryapatil) [inline astropy/timeseries/binned.py:122]: The [PEP8 conventions](https://www.python.org/dev/peps/pep-0008/#:~:text=Limit%20all%20lines%20to%20a,be%20limited%20to%2072%20characters) mention that you should limit lines to 79 characters and docstrings/comments to 72.

```suggestion
        # If any one of the parameters 'time_bin_start' or 'time_bin_end'
        # is not scalar, an exception is raised.
```

File: astropy/timeseries/binned.py
Comment (aaryapatil) [inline astropy/timeseries/binned.py:126]: The above comment applies here too.
```suggestion
                raise TypeError("'n_bins' is  not valid as 'time_bin_start' or"
                                             'time_bin_size' are not scalar.")
```

File: astropy/timeseries/binned.py
Comment (aaryapatil) [inline astropy/timeseries/binned.py:124]: Docstring at https://docs.astropy.org/en/latest/api/astropy.timeseries.BinnedTimeSeries.html does, "This is only used if both `time_bin_start` and `time_bin_size` are provided and are scalar values." The problem here is that you are checking `time_bin_end` instead of `time_bin_size`. Hopefully this solves the failing test case.

```suggestion
        if n_bins is not None and time_bin_size is not None:
            if not (time_bin_start.isscalar and  time_bin_size.isscalar):
```

**Comment by aaryapatil** [comment]
All the failing tests seem to be unrelated. The towncrier changelog error is related to the recent changes made in terms of adding logs of changes to `astropy`. I am new to towncrier too, so I'll check that myself or ask someone to help!

**Comment by aaryapatil** [comment]
You should now add a `change log` to document the change you’ve made in the astropy log. The way to do it is described [here](https://github.com/astropy/astropy/blob/main/docs/changes/README.rst). Once you do that, I’ll ask a maintainer of `timeseries` to review!

**Comment by aaryapatil** [comment]
Thanks; it looks like this is at a good stage! @astrofrog it would be great if you can take a look at this PR 😄 . You had commented in #11262, that raising an exception in this case is reasonable.

File: astropy/timeseries/binned.py
Comment (aaryapatil) [inline astropy/timeseries/binned.py:336]: I wonder if this condition is ever satisfied since `time_bin_start` and `time_bin_size` are read from columns with the start time and bin size for each bin; unless there is a single element in the table (?). If this condition is never true, it could be removed. I am not entirely sure about this so we will wait for a review.

Comment (bhavyakh) [reply astropy/timeseries/binned.py:336]: Whenever an exception is raised (or there's a need to call default table readers), the return statement always passes `n_bins` as a parameter. So in some cases (for example `function test_read()` in tests/test_binned.py and test case in astropy/docs/timeseries/io.rst ) the exception we created is raised. So either changing the test case or changing this seemed like the correct option.

In simple words, I checked that this condition is sometimes true for the given test cases, and removing it will lead to failure of the above-mentioned test cases.

Comment (aaryapatil) [reply astropy/timeseries/binned.py:336]: I agree that this change is required to fix issues with the test cases as well as more generally since we do not want to pass `n_bins = len(table)` when `time_bin_start` or `time_bin_end` is not scalar. My comment was whether we ever _need_ to pass `n_bins = len(table)`, that is, is the condition `if time_bin_start.isscalar and time_bin_size.isscalar` ever true in the case of reading a table whose columns specify `time_bin_start` and `time_bin_size`. The condition might _only_ be true when there is one element. However, I do not know enough about `timeseries` to say for sure. I will let @astrofrog decide 😄

File: astropy/timeseries/tests/test_binned.py
Comment (aaryapatil) [inline astropy/timeseries/tests/test_binned.py:94]: "things crash" is commonly understood as the interpreter raising a segmentation fault or other more "severe" issues. Maybe change this as:

```suggestion
    # Make sure an exception is raised when n_bins is passed as an argument while
```
P.S. Please don't be overwhelmed by the number of nitpicky comments. This PR is basically ready, but making sure that the code looks "good" is important.

**Review by aaryapatil — APPROVED** [review]
This looks good to me so I am approving it. This will need a maintainer's approval though, so will have to wait 😄

**Review by astrofrog — DISMISSED** [review]
Looks good to me, just two small comments below. If these are addressed my review can be dismissed and this can be merged by another maintainer (I will not be able to re-review tomorrow)

**Comment by bhavyakh** [comment]
Made the requested changes and squashed the commits!

File: astropy/timeseries/binned.py
Comment (hamogu) [reply astropy/timeseries/binned.py:334]: There may be better ways to address the root cause (why is n_bins passed in that case?), but I think this is an acceptable solution. All I'm asking for is to put a comment in the code to explain this (essentially what you said above: "When XXX happens, n_bins is passed, but that causes exception X, so here we do this ..."). 
After that, I'm happy to approve an merge.

**Review by hamogu — CHANGES_REQUESTED** [review]

**Review by hamogu — APPROVED** [review]
Oh, I see that @astrofrog already said it's good. So, let get this in!

**Pull request merged**
